### ISSUE TITLE
The current version may still have problems when multiple interfaces are present

### ISSUE THREAD

**Issue by webarchymeta** [mafintosh/multicast-dns#43]
There are cases in which a host has multiple interfaces (e.g., the host is a host for virtual machines, etc.). However the `bind` and `addMembership` API has different semantics according the to manual:

1) for `bind`:
_For UDP sockets, ... If address is not specified, the operating system will attempt to listen on **all addresses**. Once binding is complete, a 'listening' event is emitted and the optional callback function is called._
2) for `addMembership`:
_... If the multicastInterface argument is not specified, the operating **system will choose one interface** and will add membership to it. To add membership to every available interface, call addMembership **multiple times, once per interface**._

Therefore, due to this semantics mismatch, maybe the current `listening` handler:
```
    socket.on('listening', function () {
        if (!port) port = me.port = socket.address().port
        if (opts.multicast !== false) {
            try {
                socket.addMembership(ip, opts.interface);
            } catch (err) {
                that.emit('error', err)
            }
            socket.setMulticastTTL(opts.ttl || 255)
            socket.setMulticastLoopback(opts.loopback !== false)
        }
    })
```
Should be changed into something like:
```
    socket.on('listening', function () {
        if (!port) port = me.port = socket.address().port
        if (opts.multicast !== false) {
            try {
                if (opts.interfaces) {
                    opts.interfaces.forEach(mi => {
                        socket.addMembership(ip, mi);
                    });
                } else {
                    socket.addMembership(ip, opts.interface)
                }
            } catch (err) {
                that.emit('error', err)
            }
            socket.setMulticastTTL(opts.ttl || 255)
            socket.setMulticastLoopback(opts.loopback !== false)
        }
    })
```
where `opts.interfaces` can be used to specify a list of interfaces that the current instance wish to listen to and keep the `opts.interface` null (or undefined) when multiple interfaces are to be handled ?

(see also #41)

My suggestion haven't been tested in an active project, but it will be in a few days ...

**Comment by garci66** [comment]
There is an additional issue I just came accross when using multiple nics. You can't bind to 0.0.0.0:5353 and expect to receive the multicast traffic coming on the second nic. You need the call to bind to listen to 224.0.0.251.5353 instead. In my case (a quick and dirty test), I use the following:

`
  var bind = thunky(function (cb) {
    if (!port) return cb(null)
    socket.once('error', cb)
    socket.bind(port, ip, function () {
      socket.removeListener('error', cb)
      cb(null)
    })
  })
`

Otherwise, I was never receiving the packets in the process.
I belive regardless we should be binding to the group and not the 0.0.0.0 socket.

**Comment by garci66** [comment]
Looking at #38 it addresses this by keeping the bind call to just the port allways, regardless of number of interfaces specificied. I think this is nicer, especially since it will also catch unicast queries / replies.

**Comment by bnielsen1965** [comment]
> You need the call to bind to listen to 224.0.0.251.5353 instead

You may run into problems if you try this on a Windows operating system. I was using the method of binding to the multicast DNS IP address and it worked fine under linux but I ran into problems when using the code under Windows. I don't know if it is a Windows specific issue or if there was some conflict with some other application.

I should probably take pull request #42 from @panta and update the code for another PR, but for now I've reworked the code from @mafintosh into [mdns-server](https://www.npmjs.com/package/mdns-server) (I know, shameless plug ;) and I've focused my efforts there for now.

**Comment by garci66** [comment]
Indeed regarding the multicast bind. It would also block you from receving unicast queries which the protocol supports. I've gone back to basically removing the interface's ip on the call to "bind" and be done with it. I will take a look at your code as well. Not sure what the difference is.

**Comment by webarchymeta** [comment]
Yeah, binding to group ip and port, which is recommended [in a StackOverflow discussion](https://stackoverflow.com/questions/10692956/what-does-it-mean-to-bind-a-multicast-udp-socket), only works on non-Windows systems (linux, macos) for me as well.

**Comment by webarchymeta** [comment]
I have created my own fork that summarizes and incorporate all info here and on other branches. I am not sure if I should send a pull request since I also changed the coding style (to ES6) and the name an items in the options ...

**Comment by mafintosh** [comment]
Just pushed a new minor that should have much better support for this. Feedback welcome

**Comment by mafintosh** [comment]
If this is still not fixed, please re-open
